### PR TITLE
ci: sync Markdownlint setup from mdn/content

### PR DESCRIPTION
**Pull request by nschonni**
Copied the setup from the upstream repo, but added a step to copy over the package.json and yarn.lock to use the same version of the tools

### CODE REVIEW

**Comment by nschonni** [comment]
Note that some of the issues will be silenced with https://github.com/mdn/translated-content/pull/7778, and normally the second Markdownlint that only runs on changed markdown files will be run.

/cc @caugner @yin1999 @OnkarRuikar

File: .github/workflows/markdown-lint.yml
Comment (yin1999) [inline .github/workflows/markdown-lint.yml:22]: should this be:

```suggestion
          curl -L https://raw.githubusercontent.com/mdn/content/main/package.json > package.json
          curl -L https://raw.githubusercontent.com/mdn/content/main/yarn.lock > yarn.lock
```

Comment (nschonni) [reply .github/workflows/markdown-lint.yml]: Thanks, that got left over from the original try with `-O`. I rebased to fix it in both files

File: .github/workflows/pr-check_markdownlint.yml
Comment (OnkarRuikar) [inline .github/workflows/pr-check_markdownlint.yml:26]:
```suggestion
          curl -L https://raw.githubusercontent.com/mdn/content/main/package.json -o package.json
```
file can be saved in the same command

File: .github/workflows/pr-check_markdownlint.yml
Comment (OnkarRuikar) [inline .github/workflows/pr-check_markdownlint.yml:27]:
```suggestion
          curl -L https://raw.githubusercontent.com/mdn/content/main/yarn.lock -o yarn.lock
```

**Review by OnkarRuikar — CHANGES_REQUESTED** [review]

**Comment by OnkarRuikar** [comment]
We shouldn't link the linter version to the `mdn/content` project. Because there version gets updated after all the files are fixed.
Here also we need to make sure all the file are fixed before changing the linter version. The two project are not linked in this regard.
***
We do not have to install all the content/yari dependancies to lint the translated content:
```
      - uses: actions/checkout@v3

      - name: Setup Node.js environment
        uses: actions/setup-node@v3
        with:
          node-version: "16"

      - name: Setup packages
        run: |
          npm init -y &> /dev/null
          version=$( curl https://raw.githubusercontent.com/mdn/content/main/package.json 2>&1 | grep -oP "(?<=markdownlint-cli....)[\.\d]+" )
          npm install -g "markdownlint-cli@$version"

      - name: Lint markdown files
        run: |
          echo "::add-matcher::.github/workflows/markdownlint-problem-matcher.json"
          markdownlint "**/*.md"
```

**Comment by caugner** [comment]
> We shouldn't link the linter version to the `mdn/content` project.

Very valid point from @OnkarRuikar, so how about having a workflow instead that opens a PR with these changes?

**Comment by nschonni** [comment]
> We shouldn't link the linter version to the `mdn/content` project. Because there version gets updated after all the files are fixed.
> Here also we need to make sure all the file are fixed before changing the linter version. The two projects are not linked in this regard.

They content is synced across, so I think I'm not sure why they wouldn't use the same tools. This repo hasn't been cleaned up and is still in the process of converting to Markdown, so having the task now would help flag stuff to cleanup as it lands.

If you think it shouldn't leverage the lock/package from the other project, I think there should be a separate copy in this repo+dependabot rather than embedded versions in the action files

**Comment by OnkarRuikar** [comment]
> so I think I'm not sure why they wouldn't use the same tools.

It's not about the tools being same. It's about tool versions. The content project doesn't wait for this repo to update the tool versions.

> If you think it shouldn't leverage the lock/package from the other project, I think there should be a separate copy in this repo+dependabot rather than embedded versions in the action files

I agree. We need to do `npm init` here and add only the required tools. And the dependabot.
To avoid accidental publish/release, set `"private": true` and avoid `name` and `version` props in the package.json. [[ref](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#name)].
For example, the [mdast repo setup](https://github.com/syntax-tree/mdast/blob/main/package.json)

**Comment by nschonni** [comment]
I ran `npm init --y` and cleaned it up a bit. Used Yarn to match the tooling upstream

File: .gitignore
Comment (OnkarRuikar) [inline .gitignore:3]:
```suggestion
node_modules/
yarn-debug.log*
yarn-error.log*
```

File: package.json
Comment (OnkarRuikar) [inline package.json:7]:
```suggestion
    "lint:md": "markdownlint \"**/*.md\" -i node_modules",
    "fix:md": "npm run lint:md -- --fix"
```
allow contributors to fix errors in local before commits

File: package.json
Comment (OnkarRuikar) [inline package.json:14]:
```suggestion
  "author": "MDN Web Docs",
```
if we are providing this field then it could be like what yari has

File: package.json
Comment (OnkarRuikar) [inline package.json:15]:
```suggestion
  "license": "SEE LICENSE IN LICENSE.md",
```
to mimic content

File: package.json
Comment (OnkarRuikar) [inline package.json:13]:
```suggestion
```
since this is not going to get published we don't need keywords

**Comment by OnkarRuikar** [comment]
In order for contributors to avoid facing errors in local environment we need to provide engine version in package.json:
The `markdownlint-cli` supports engines >=14:
```json
"engines": {
    "node": ">=14"
  },
```
We can have the same or as we are starting a fresh we can start with the latest engine `>=18.0.0` (like yari)?

**Comment by nschonni** [comment]
I think CI is mostly 16 right now, so 14/16 would probably be better for now if you want to add the engine

**Comment by caugner** [comment]
> I think CI is mostly 16 right now, so 14/16 would probably be better for now if you want to add the engine

Let's use the same version constraint as yari and content (after https://github.com/mdn/content/pull/20240).

**Comment by nschonni** [comment]
I've add the engines to match the pending PR upstream, and added the MD051/52 disable to this PR (which knocks off 3K issues)

**Comment by nschonni** [comment]
@caugner I've got a bunch of PRs to cleanup a bunch of the issues, but this should be good to land

File: package.json
Comment (caugner) [inline package.json:2]:
```suggestion
  "name": "@mdn/translated-content",
```

File: .github/workflows/pr-check_markdownlint.yml
Comment (caugner) [inline .github/workflows/pr-check_markdownlint.yml:19]: I feel like pinning actions to the major version is sufficient:

```suggestion
        uses: tj-actions/changed-files@v29
```

File: .github/workflows/pr-check_markdownlint.yml
Comment (caugner) [inline .github/workflows/pr-check_markdownlint.yml:32]: This command can be moved to the previous line without the vertical dash.

**Review by caugner — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-check_markdownlint.yml
Comment (nschonni) [reply .github/workflows/pr-check_markdownlint.yml:32]: Agreed, but I left it like the others/upstream

File: package.json
Comment (nschonni) [reply package.json]: This one is marked as private, but doesn't hurt to have it in better shape if that ever changes

File: .github/workflows/pr-check_markdownlint.yml
Comment (nschonni) [reply .github/workflows/pr-check_markdownlint.yml]: Agreed, you might want to do the same upstream

**Review by caugner — APPROVED** [review]
LGTM

**Comment by nschonni** [comment]
This can either land as-is, or I'll updated it to match https://github.com/mdn/content/pull/20330 if that lands

**Comment by yin1999** [comment]
> This can either land as-is, or I'll updated it to match [mdn/content#20330](https://github.com/mdn/content/pull/20330) if that lands

I'm for updating to markdownlint-cli2.

**Comment by nschonni** [comment]
Flipping to draft to switch this over to CLI2, now that the upstream PR landed

**Comment by nschonni** [comment]
OK, I've migrated this over to CLI2 to match upstream

File: .markdownlint-cli2.jsonc
Comment (caugner) [inline .markdownlint-cli2.jsonc:126]: Missing trailing newline

**Review by caugner — APPROVED** [review]
LGTM, just one nit.

File: .markdownlint-cli2.jsonc
Comment (nschonni) [reply .markdownlint-cli2.jsonc]: Thanks, copied the `.editorconfig` from mdn/content and resaved

**Comment by nschonni** [comment]
@caugner could we land this one?

**Comment by caugner** [comment]
Given the amount of outstanding errors (8811), I feel like this needs to be discussed with @mdn/localization-team-leads 

Here's an overview of the errors:

```console
$ gh pr checkout 8181
$ yarn install
$ yarn lint:md 2> lint.log
$ rg -iNo 'MD\d+/[a-z-]+' lint.log | sort | uniq -c | sort -rn                                                                                                                                                                                                              ✭
6755 MD010/no-hard-tabs
1663 MD033/no-inline-html
 111 MD046/code-block-style
 110 MD030/list-marker-space
  78 MD038/no-space-in-code
  32 MD039/no-space-in-links
  26 MD012/no-multiple-blanks
  15 MD032/blanks-around-lists
  10 MD009/no-trailing-spaces
   5 MD031/blanks-around-fences
   2 MD011/no-reversed-links
   2 MD003/heading-style
   1 MD047/single-trailing-newline
   1 MD022/blanks-around-headings
```

French locale uses `<i lang="en"></i>` to mark English words, and I'm not sure how this can be achieved without inline html (MD033), so that rule should probably be disabled.

**Review by caugner — CHANGES_REQUESTED** [review]
Disable MD033 (unless an alternative exists), and wait for feedback from localization team leads.

**Comment by nschonni** [comment]
Fine to wait on the leads. I did ignore and comment out a few more of the tags for MD033.
I believe the bulk of those open issues are addressed by the 3 RU cleanup PRs I've got open.
There are a few errors that exist here from older format that doen't exist in content repo, so I didn't include the CI job to try and auto fix things yet. Right now this is just to start flagging new stuff during the conversion process.

**Comment by nschonni** [comment]
Realized the "conflicting" and "orphaned" folders under the languages should probably be skipped, so I added them to the ignore paths.

File: .github/workflows/pr-check_markdownlint.yml
Comment (OnkarRuikar) [inline .github/workflows/pr-check_markdownlint.yml:37]:
```suggestion
          echo "::add-matcher::.github/workflows/markdownlint-problem-matcher.json"
          files_to_lint="${{ steps.changed-files.outputs.all_changed_files }}"
          yarn markdownlint-cli2 ${files_to_lint}
```
There are 66 files in the repo with `(` in the path.

**Comment by SphinxKnight** [comment]
I need to spend more time on this one but: 

- What happens for docs which are not yet in Markdown? Once they're converted, do we expect them to play nice and pretty or to have a bunch of linting errors (will this block any conversion PR from being merged?)
- I'll check the `ru` PRs and if they're waiting for more than a month, will merge them.

**Comment by nschonni** [comment]
I could also add the RU docs to the the ignore section, if they're not that active currently

**Comment by YangFong** [comment]
Please refer to mdn/content#20879 to update `.gitignore`.

**Comment by nschonni** [comment]
I've seen that ignored in some repos, but I think that one is better ignored in your own machine .gitignore, unless a particular IDE is part of the common toolset for a project

**Comment by nschonni** [comment]
Also not really related to adding Markdownlint to this repo

**Comment by YangFong** [comment]
Maybe I should have initiated the PR alone instead of doing it here.

**Comment by schalkneethling** [comment]
Hey @nschonni, @OnkarRuikar, and @caugner, how about we "abstract" this to the `mdn/workflows` repo and then reuse these workflows across repos? We can make them configurable so each repository can change some of the defaults as needed. Thoughts?

**Comment by nschonni** [comment]
> Hey @nschonni, @OnkarRuikar, and @caugner, how about we "abstract" this to the `mdn/workflows` repo and then reuse these workflows across repos? We can make them configurable so each repository can change some of the defaults as needed. Thoughts?

Some of the PR check linting jobs probably could be useful across the repos and could be genericized, although the steps are fairly simple.
I'm not sure how many will leverage the Markdownlint ones though, outside of these 2, but it could always be hoisted there if more look like they could leverage them later

**Comment by nschonni** [comment]
I've made a few PRs in workflows repos, but haven't aligned the markdownlint ones fully to what's in here yet.

**Comment by nschonni** [comment]
Gave this a rebase and added the extra job to allow separate auto-fix jobs for each locale. That will leave the root files ignored for now, but those files probably get touched less, and are less likely to need that for now

**Comment by caugner** [comment]
@nschonni Can you take a look at those conflicts? I will merge on Monday if all looks good.

**Comment by nschonni** [comment]
@Graywolf9 @SphinxKnight since I think you are doing the heavy lifting on a bunch of the remaining HTML -> MD, would you prefer to wait till those PRs land first, or are you OK with this landing?

**Comment by SphinxKnight** [comment]
> @Graywolf9 @SphinxKnight since I think you are doing the heavy lifting on a bunch of the remaining HTML -> MD, would you prefer to wait till those PRs land first, or are you OK with this landing?

As for `ru`, the last conversion PRs should be merged by the next 10 days (it'll be ~1 month since I opened them). I'd be thankful to wait a bit.

I'll let @mdn/yari-content-es speak for `es`.

**Comment by nschonni** [comment]
Copied over the lint-staged/husky setup to help autofix minor issues on commit, but it will block the commit on un-fixable things. I can pull this out for a later PR though

**Comment by nschonni** [comment]
@Graywolf9 @SphinxKnight looks like the conversion is done. I've got one open PR for a chunk for the RU cleanup and i'll try and add the rest. Do you think this is OK to land now?

**Review by yin1999 — APPROVED** [review]

**Comment by SphinxKnight** [comment]
@nschonni thanks again for the hard work on this and your patience :bow: 

Let's go and land this :) 
Sign-off from me :)

**Review by SphinxKnight — APPROVED** [review]

**Comment by nschonni** [comment]
@caugner down to one minor space issue with an open PR, but that could also be used for the autofix PR job

File: .github/workflows/markdown-lint-fix.yml
Comment (OnkarRuikar) [inline .github/workflows/markdown-lint-fix.yml:32]:
```suggestion
      - name: Install all yarn packages
        env:
          HUSKY: 0
```

**Comment by nschonni** [comment]
@caugner would be good to get this one landed now that the conversion is done

**Comment by nschonni** [comment]
@schalkneethling not sure if you can land this one with the one blocking review from @caugner or not

**Review by caugner — APPROVED** [review]

**Comment by nschonni** [comment]
Ping @caugner it would be good to land so we can see if the autofix PRs work. It's OK to have the job failure right now

**Comment by nschonni** [comment]
@caugner do you want me to rebase/squash the commits?

**Comment by Rumyra** [comment]
Yes could you @nschonni - I'll see if we can get final review and merge 👍

**Comment by bsmth** [comment]
Also having a look 👀

File: .github/workflows/pr-check_markdownlint.yml
Comment (bsmth) [inline .github/workflows/pr-check_markdownlint.yml:19]: Not a blocker, but this is [up to `v34.4.4` now](https://github.com/tj-actions/changed-files/releases/tag/v34.4.4), there were some bugs we hit often on the content repo that are now fixed in later versions.

File: .markdownlint-cli2.jsonc
Comment (bsmth) [inline .markdownlint-cli2.jsonc:25]:
```suggestion
    // Consecutive Notes/Callouts currently don't conform with this rule
```

**Review by bsmth — APPROVED** [review]
<strike>One comment open about using latest GH action, but can be addressed in a follow-up.</strike> 

__Edit:__ the gh action is now bumped to `tj-actions/changed-files@v34`

I don't see anything wrong (aside from failing CI 😄), so also leaving a LGTM 👍🏻

File: .github/workflows/pr-check_markdownlint.yml
Comment (nschonni) [reply .github/workflows/pr-check_markdownlint.yml]: Updated to the floating `v34`

**Comment by nschonni** [comment]
The failures for the job can be used to test the Autofix PR job

**Comment by nschonni** [comment]
There is an odd fix for the headings in blockquote that I'm going to raise upstream, but not blocking this

File: .github/workflows/pr-check_markdownlint.yml
Comment (bsmth) [reply .github/workflows/pr-check_markdownlint.yml]: Thank you! Is this equivalent to `v34@latest`? We had issues with `v34.4.2` as far as I remember.

Comment (bsmth) [reply .github/workflows/pr-check_markdownlint.yml]: OK understood, thanks 👍🏻

**Comment by schalkneethling** [comment]
> There is an odd fix for the headings in blockquote that I'm going to raise upstream, but not blocking this

Are we OK with Markdownlint failing? Looking at the log there are no "errors" but there are 20 warnings.

**Comment by nschonni** [comment]
Upstream, they changed it from Errors to Warning to the severity. I'm not sure I get the benefit, as the job will fail either way, but I guess it looks less scary on the reviewers.

The 20 issues are fine, as this PR includes a new job that will create PRs to fix the existing issues. I created a manual one for the [Japanese texts](https://github.com/mdn/translated-content/pull/10121) since there are formatting issues that the autofix won't quite handle. Landing this is fine though

**Comment by OnkarRuikar** [comment]
> Are we OK with Markdownlint failing? Looking at the log there are no "errors" but there are 20 warnings.

Yes. [We don't want to burden online(on Github) contributors](https://github.com/mdn/content/pull/12316#issuecomment-1219144778) with lint errors, cos they may not know Markdown syntax, so the lint errors are marked as warning. Also because we have [a bot in place that daily autofixes the lint issues](https://github.com/mdn/content/pulls?q=Markdownlint+auto-cleanup).
We purposefully fail the job on warnings because it shows the indicator in PR list. And some maintainers proactively look for the failing test PRs and fix the lint issues.

Note: Fixing lint issues has been made mandatory for those who push changes using git.

**Comment by nschonni** [comment]
Thanks @caugner ! Now to wait and see how the auto-PRs go 😄

**Comment by yin1999** [comment]
Sorry for ignoring the PAT field, should we use `AUTOMERGE_TOKEN` in the markdownlint-fix workflow?

https://github.com/mdn/translated-content/blob/e0c9430fd9190dfbebe5917129505527cbdf7e3b/.github/workflows/sync-translated-content.yml#L67

**Comment by nschonni** [comment]
Yeah, I guess that would be needed to have the PRs trigger subsequent CI jobs

**Comment by hochan222** [comment]
Recently, while conducting a review, I encountered a lot of errors like the following. If the contributor re-commits, it will be resolved(There are times when it isn't.), but I'm worried about what instructions to give. Or.. I am concerned about github actions that did not pass through ignoring and merging.

Of course, @nschonni replied to me, but I'll ask again to confirm. Should I tell the contributor to fix it?

> Error: Unable to locate a common ancestor between main and c1c4e5ca2a6a3fee05f603f1daa36bb83cda2157

PR:  https://github.com/mdn/translated-content/pull/10063#pullrequestreview-1195076245

**Comment by OnkarRuikar** [comment]
@hochan222 The error is not related to the changes done in that PR; it's related to the workflow code. Try [re-running the workflow](https://github.com/orgs/community/discussions/25726#discussioncomment-3248945). Go to the failing check's log and click on Re-run jobs -> Re-run all jobs.

If that fails rebase the PR's branch, which will automatically trigger all the workflows..

***
[Linting is should not mandatory](https://github.com/mdn/translated-content/pull/8181#issuecomment-1325871051). I suggest simply merge the PR. If there are linting errors, they will be fixed by cron job next midnight UTC.

**Comment by hochan222** [comment]
@OnkarRuikar Thank you for your reply. The error I encountered in various PRs was that the markdownlint workflow failed after further changes, even though the first time passed fine. Even running the workflow again failed the same. In the end, I asked contributors to rebase or commit for this.

image

Also, In the way to merge right away... the reason I wrote it for confirmation was because I was concerned about that it was displayed as an **X** ​​rather than a check..

**Comment by hochan222** [comment]
I merged PRs that failed the workflow, but the check went well as shown below. I think just merging would be fine.



image

**Pull request merged**
